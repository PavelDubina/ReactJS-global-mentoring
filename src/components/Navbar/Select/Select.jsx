import React, { useState } from 'react'
import className from 'classnames'
import { sortingValues } from '../../../utils/constants'
import styles from './Select.scss'

export const Select = () => {
  const [isOpen, setIsOpen] = useState(false)
  const [value, setValue] = useState('RELEASE DATE')
  const selectStyle = className(styles.select, { [styles.open]: isOpen })
  const dropdownStyle = className(styles.dropdown, { [styles.visible]: isOpen })
  return (
    <div onClick={() => setIsOpen((state) => !state)} className={selectStyle}>
      {value}
      <div onClick={(event) => setValue(event.target.textContent)} className={dropdownStyle}>
        {sortingValues.map((val) => {
          const activeStyle = className({ [styles.active]: val === value })
          return (
            <p className={activeStyle} key={val}>
              {val}
            </p>
          )
        })}
      </div>
    </div>
  )
}
