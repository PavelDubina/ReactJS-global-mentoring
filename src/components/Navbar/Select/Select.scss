.select {
  position: relative;
  height: 24px;
  width: 160px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin: 0 0 0 24px;
  color: #fff;
  cursor: pointer;

  &::after {
    content: '\25BC';
    margin: 0 0 0 10px;
    font-size: 12px;
    color: #f65261;
  }
}

.open {
  &::after {
    transform: rotate(180deg);
  }
}

.dropdown {
  position: absolute;
  display: none;
  right: 0;
  top: 26px;
  width: 174px;
  background-color: #232323;
  border: 1px solid #424242;
  border-radius: 4px;
  z-index: 2;

  & p {
    line-height: 32px;
    padding: 0 8px;
  }
}

.visible {
  display: block;
}

.active {
  background-color: #f65261;
}
