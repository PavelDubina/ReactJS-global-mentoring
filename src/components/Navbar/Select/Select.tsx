import React, { useState, useRef } from 'react'
import className from 'classnames'
import { useTranslation } from 'next-i18next'
import { useOutsideClick } from '../../../Hooks/useClickOutside'
import { sortingValues } from '../../../utils/constants'
import styles from '../../../../styles/Select.module.scss'
import { getSortQuery } from '../../../utils/helpers'

type SelectProps = {
  value: string
  toggleSortValue: (sortValue: string) => void
}

export const Select: React.FC<SelectProps> = ({ value, toggleSortValue }) => {
  const { t } = useTranslation()
  const ref = useRef(null)
  const [isOpen, setIsOpen] = useState(false)
  const selectStyle = className(styles.select, { [styles.open]: isOpen })
  const dropdownStyle = className(styles.dropdown, { [styles.visible]: isOpen })
  useOutsideClick(ref, () => isOpen && setIsOpen(false))
  return (
    <div data-testid="select" onClick={() => setIsOpen((state) => !state)} ref={ref} className={selectStyle}>
      {t(`nav.sort.${getSortQuery(value)}`)}
      <div data-testid="dropdown" className={dropdownStyle}>
        {Object.values(sortingValues).map((val) => {
          const activeStyle = className({ [styles.active]: val === value })
          return (
            <p onClick={() => toggleSortValue(val)} data-testid={val} className={activeStyle} key={val}>
              {t(`nav.sort.${getSortQuery(val)}`)}
            </p>
          )
        })}
      </div>
    </div>
  )
}
