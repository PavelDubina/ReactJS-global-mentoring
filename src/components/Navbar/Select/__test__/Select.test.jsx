import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { Select } from '../Select'

describe('Select', () => {
  const getComponent = ({ value = 'RELEASE DATE', toggleSortValue = jest.fn() }) =>
    render(<Select value={value} toggleSortValue={toggleSortValue} />)
  it('Should renders correctly', () => {
    expect(getComponent({}).asFragment()).toMatchSnapshot()
  })
  it('Dropdown should be visible', () => {
    const { getByTestId } = getComponent({})
    expect(getByTestId('dropdown').classList.contains('visible')).toBeFalsy()
    fireEvent.click(getByTestId('select'))
    expect(getByTestId('dropdown').classList.contains('visible')).toBeTruthy()
    fireEvent.click(document)
    expect(getByTestId('dropdown').classList.contains('visible')).toBeFalsy()
  })
})
