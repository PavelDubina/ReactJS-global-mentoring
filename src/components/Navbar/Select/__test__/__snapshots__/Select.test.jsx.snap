// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Should renders correctly 1`] = `
<DocumentFragment>
  <div
    class="select"
    data-testid="select"
  >
    RELEASE DATE
    <div
      class="dropdown"
      data-testid="dropdown"
    >
      <p
        class="active"
        data-testid="RELEASE DATE"
      >
        RELEASE DATE
      </p>
      <p
        class=""
        data-testid="RATING"
      >
        RATING
      </p>
    </div>
  </div>
</DocumentFragment>
`;
